### ISSUE TITLE
Comments out certain markup during hydration in latest version

### ISSUE THREAD

**Issue by ljosberinn** [preactjs/preact#2950]
Hi, just noticed that in the latest version. Repo is Next.js; you can see its correct server-side as its shortly visible before hydration kicks in.

### Reproduction

[PR for 10.5.10](https://github.com/ljosberinn/gerritalex.de/pull/381) - [deployment](https://gerritalex-de-iw7iebtq3.vercel.app/)

![image](https://user-images.githubusercontent.com/29307652/105344344-18fefd00-5be3-11eb-8fb0-0abf28649a16.png)

[PR for 10.5.11](https://github.com/ljosberinn/gerritalex.de/pull/385) - [deployment](https://gerritalex-de-gv62vjwsx.vercel.app/)

![image](https://user-images.githubusercontent.com/29307652/105344364-1e5c4780-5be3-11eb-8829-ca77b2e88369.png)

![image](https://user-images.githubusercontent.com/29307652/105344406-2a480980-5be3-11eb-8b55-ea7aecba3177.png)

[source](https://github.com/ljosberinn/gerritalex.de/blob/b1ec0e6ec6267def58dbb3f3e02d45e1888b7dc6/components/Sites.js#L91)

between those, only dev dependencies were merged
![image](https://user-images.githubusercontent.com/29307652/105344305-0d133b00-5be3-11eb-818e-16dc4b810ddc.png)

### Expected Behavior

No diff between server and client.

### Actual Behavior

Client behaves differently.

**Comment by kspeyanski** [comment]
Hello, 

I'm also experiencing this, but in `gatsby` project. 

Please let me know if i can provide any additional information to help resolving this issue.

**Comment by marvinhagemeister** [comment]
A short codesandbox or repo would be a godsend. If we can reproduce errors on our machines it's much easier for us to get to the bottom of it 👍

**Comment by ljosberinn** [comment]
Sorry, was at work! There you go. 

https://github.com/ljosberinn/preact-hydration-bug
https://preact-hydration-bug.vercel.app/

**Comment by marvinhagemeister** [comment]
@ljosberinn Awesome, thank you so much for the repro case. With it I was able to narrow it down to the server side rendered HTML not being generated by `preact-render-to-string`. Instead the generated output comes from `react-dom/server`, whose output differs slightly enough by including comment nodes in the HTML that in turn trip up our reconciler.

A fix for this is to replace the custom `withPreact()` function in `next.config.js` with our official plugin for nextjs [next-plugin-preact](https://github.com/preactjs/next-plugin-preact):

```diff
+ const withPreact = require("next-plugin-preact");
- const withPreact = (config, options) => {
-   if (!options.dev) {
-     const splitChunks = config.optimization && config.optimization.splitChunks;

-     if (splitChunks) {
-       const { cacheGroups } = splitChunks;
-       const test = /[\\/]node_modules[\\/](preact|preact-render-to-string|preact-context-provider)[\\/]/u;
-       if (cacheGroups.framework) {
-         cacheGroups.preact = {
-           ...cacheGroups.framework,
-           test,
-         };

-         cacheGroups.commons.name = "framework";
-       } else {
-         cacheGroups.preact = {
-           chunks: "all",
-           name: "commons",
-           test,
-         };
-       }
-     }

-     const aliases = config.resolve.alias || (config.resolve.alias = {});
-     aliases.react = aliases["react-dom"] = "preact/compat";
-     aliases["react-ssr-prepass"] = "preact-ssr-prepass";
-   }
- };

- module.exports = {
-   webpack: (config, options) => {
-     withPreact(config, options);
-
-     return config;
-   },
- };
+ module.exports = withPreact({
+   // other custom nextjs config here
+ });
```

With that in place the issue is resolved for me:

<img width="335" alt="Screenshot 2021-01-22 at 06 52 22" src="https://user-images.githubusercontent.com/1062408/105452309-63818780-5c7e-11eb-9845-c222f7e3224f.png">

**Comment by marvinhagemeister** [comment]
@kspeyanski Can you file a separate issue? They are likely not related, even though the symptoms may look similar.

**Comment by kspeyanski** [comment]
Hello @marvinhagemeister ,

I've created the following project where the exact same issue (commenting out certain markup after hydratation) is reproducible: please find it here: https://github.com/kspeyanski/preact-hydratation-bug

I still suspect that the `gatsby` and `nextjs` issues are related as both:
- results in the exact same markup
- are only reproducible with version `10.5.11` of `preact`

Please further advice if a separate issue is needed, as at this point i would have to give it an exact same title as this one, and i would like to avoid spam.

**Comment by marvinhagemeister** [comment]
@kspeyanski Thanks for the repro! Just took a look and you're right in that it is caused by the same issue as in the next plugin. They are using React to render a virtual tree built by Preact on the server. I'm kinda surprised that this setup worked at all, but then again it's sort of a good testament to how strong our compatibility layer is.

To resolve the issue in your repo, instead this additional packages:

```bash
yarn add react@npm:@preact/compat react-dom@npm:@preact/compat
```

**Comment by kspeyanski** [comment]
Thank you @marvinhagemeister for the fast response.

However, my specific use case does not allow for installing the packages through aliasing, thus i'm resolving the issue by downgrading to `v10.5.10` for now.

Feel free to reach me out in case you need more information.

**Comment by marvinhagemeister** [comment]
@wardpeet Do you know of a way to swap out the server renderer in Gatsby via a plugin?

**Comment by wardpeet** [comment]
That should be with gatsby-plugin-preact, I'll try the reproduction and fix it in the plugin. Thanks for the ping!

**Comment by marvinhagemeister** [comment]
FYI: We just released 10.5.12 which reverts that change that lead to this issue :+1:
